### PR TITLE
disttask: do not view normal nodes as background nodes when all background nodes fail

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/50538

Problem Summary:

### What changed and how does it work?

- add `dead` field to dist_framework_meta to check existed dead node

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by ywqzzy** [comment]
What will happen when dead node restart? Check the related row in dist_framework_meta.

**Comment by okJiang** [comment]
> What will happen when dead node restart? Check the related row in dist_framework_meta.

pr is not completed, one commit is later coming. After test it, I will remove the [wip] lable

**Comment by okJiang** [comment]
> What will happen when dead node restart? Check the related row in dist_framework_meta.

set dead=false when insert record on duplicate key. https://github.com/pingcap/tidb/pull/50540/files#diff-fde849081cda4cf9d6e56f8f339efb65ce401d3e444c89ac7681820be846e73cR46-R61

File: pkg/session/bootstrap.go [hunk 1]
```diff
@@ -612,6 +612,7 @@ const (
         host VARCHAR(100) NOT NULL PRIMARY KEY,
         role VARCHAR(64),
         cpu_count int default 0,
+		dead BOOL NOT NULL DEFAULT FALSE,
```
Comment (ywqzzy) [inline pkg/session/bootstrap.go:615]: fmt

File: pkg/disttask/framework/storage/nodes.go [hunk 2]
```diff
@@ -100,14 +102,26 @@ func (mgr *TaskManager) getManagedNodesWithSession(ctx context.Context, se sessi
...
+			managedNodes = append(managedNodes, node)
+		}
 	}
-	if len(nodeMap["background"]) == 0 {
-		return nodeMap[""], nil
+	if !setTiDBServiceScopeBackground {
+		return liveNodes, nil
 	}
-	return nodeMap["background"], nil
+
```
Comment (ywqzzy) [inline pkg/disttask/framework/storage/nodes.go:123]: We can do this in 2 sql

Comment (okJiang) [reply pkg/disttask/framework/storage/nodes.go:123]: Can you explain in more detail? I didn't understand what you meant.

File: pkg/disttask/framework/storage/nodes.go [hunk 3]
```diff
@@ -57,29 +58,30 @@ func (mgr *TaskManager) RecoverMeta(ctx context.Context, execID string, role str
...
-			deleteElems = append(deleteElems, fmt.Sprintf(`"%s"`, node))
+			updateElems = append(updateElems, fmt.Sprintf(`"%s"`, node))
 		}
 
-		deleteSQL.WriteString(strings.Join(deleteElems, ", "))
-		deleteSQL.WriteString(")")
-		_, err := sqlexec.ExecSQL(ctx, se, deleteSQL.String())
+		updateSQL.WriteString(strings.Join(updateElems, ", "))
+		updateSQL.WriteString(")")
+		_, err := sqlexec.ExecSQL(ctx, se, updateSQL.String())
```
[the remark this answers is not in the document]
Comment (okJiang) [reply pkg/disttask/framework/storage/nodes.go]: Did you mean change it to this?

```golang
    fmt.Sprintf(`...host in(%s)`, strings.Join(updateElems, ", ")
```

File: pkg/disttask/framework/storage/nodes.go (hunk 2 above)
Comment (ywqzzy) [reply pkg/disttask/framework/storage/nodes.go:123]: select id, role, cpu_count from meta where role="background" and dead=false;
select id, role, cpu_count from meta where role="" and dead=false;

File: pkg/session/bootstrap_test.go [hunk 4]
```diff
@@ -2183,3 +2183,32 @@ func TestTiDBUpgradeToVer179(t *testing.T) {
 
 	dom.Close()
 }
+
+func TestTiDBUpgradeToVer184(t *testing.T) {
+	store, _ := CreateStoreAndBootstrap(t)
```
Comment (D3Hunter) [inline pkg/session/bootstrap_test.go:2188]: we're in 184 already after this line

File: pkg/disttask/framework/storage/nodes.go [hunk 5]
```diff
@@ -57,29 +58,30 @@ func (mgr *TaskManager) RecoverMeta(ctx context.Context, execID string, role str
...
+func (mgr *TaskManager) MarkDeadNodes(ctx context.Context, nodes []string) error {
 	if len(nodes) == 0 {
 		return nil
 	}
 	return mgr.WithNewTxn(ctx, func(se sessionctx.Context) error {
-		deleteSQL := new(strings.Builder)
-		if err := sqlescape.FormatSQL(deleteSQL, "delete from mysql.dist_framework_meta where host in("); err != nil {
+		updateSQL := new(strings.Builder)
+		if err := sqlescape.FormatSQL(updateSQL, `update mysql.dist_framework_meta
+			set dead = 1 where host in(`); err != nil {
```
Comment (D3Hunter) [inline pkg/disttask/framework/storage/nodes.go:74]:
```suggestion
			set dead = true where host in(`); err != nil {
```

File: pkg/disttask/framework/storage/nodes.go [hunk 6]
```diff
@@ -100,14 +102,26 @@ func (mgr *TaskManager) getManagedNodesWithSession(ctx context.Context, se sessi
 	if err != nil {
 		return nil, err
 	}
-	nodeMap := make(map[string][]proto.ManagedNode, 2)
+	managedNodes := make([]proto.ManagedNode, 0)
+	liveNodes := make([]proto.ManagedNode, 0)
+	var setTiDBServiceScopeBackground bool
```
Comment (D3Hunter) [inline pkg/disttask/framework/storage/nodes.go:107]:
```suggestion
	var haveBackgroundNode bool
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/storage/nodes.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/disttask/framework/storage/nodes.go:123]: lgtm to use one, cheaper(1 round trip)

node count is relatively small normally

Comment (ywqzzy) [reply pkg/disttask/framework/storage/nodes.go:123]: OK

File: pkg/disttask/framework/storage/nodes.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ywqzzy) [reply pkg/disttask/framework/storage/nodes.go]: yes

[the remark this answers is not in the document]
Comment (okJiang) [reply pkg/disttask/framework/storage/nodes.go]: The reason that using strings.Builder, I guess, is to reduce overhead and improve performance. If the readability impact is minimal, let's stick with the current state?

https://medium.com/@teamcode20233/a-comprehensive-guide-for-golang-string-concatenation-df1a02210492

File: pkg/session/bootstrap_test.go (hunk 4 above)
Comment (okJiang) [reply pkg/session/bootstrap_test.go:2188]: Seems we do not need this ut? 🤔

**Review by D3Hunter — DISMISSED** [review]

**Comment by D3Hunter** [comment]
/hold

require discuss node manage behavior

**Comment by okJiang** [comment]
Without optimizing for the time being, it can be noted in the document that after all background nodes are unexpectedly offline, existing tasks will be scheduled to nodes with tidb_service_scope="", which may have the risk of affecting the business of other nodes. It is recommended that multiple background nodes work in the production environment.

**Pull request closed without merging**